delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/Resolvers.scala

```diff
@@ -0,0 +1,267 @@
+
+  private[delta] def next(state: ResolverState, event: ResolverEvent): ResolverState = {
+    (state, event) match {
+      // Creating an in-project resolver
+      case (Initial, e: InProjectResolverCreated)                           =>
+        e.into[InProjectCurrent]
+          .withFieldConst(_.deprecated, false)
+          .withFieldConst(_.createdAt, e.instant)
+          .withFieldConst(_.createdBy, e.subject)
+          .withFieldConst(_.updatedAt, e.instant)
+          .withFieldConst(_.updatedBy, e.subject)
+          .transform
+      // Updating an in-project resolver
+      case (c: InProjectCurrent, e: InProjectResolverUpdated)               =>
+        c.copy(priority = e.priority, rev = e.rev, updatedAt = e.instant, updatedBy = e.subject)
+
+      // Creating a cross-project resolver
+      case (Initial, e: CrossProjectResolverCreated)                        =>
+        e.into[CrossProjectCurrent]
+          .withFieldConst(_.deprecated, false)
+          .withFieldConst(_.createdAt, e.instant)
+          .withFieldConst(_.createdBy, e.subject)
+          .withFieldConst(_.updatedAt, e.instant)
+          .withFieldConst(_.updatedBy, e.subject)
+          .transform
+      // Updating a cross-project resolver
+      case (c: CrossProjectCurrent, e: CrossProjectResolverUpdated)         =>
+        c.copy(
+          priority = e.priority,
+          resourceTypes = e.resourceTypes,
+          projects = e.projects,
+          identities = e.identities,
+          rev = e.rev,
+          updatedAt = e.instant,
+          updatedBy = e.subject
+        )
+
+      // Deprecating a in-project resolver
+      case (c: InProjectCurrent, e: ResolverDeprecated) if !c.deprecated    =>
+        c.copy(rev = e.rev, deprecated = true, updatedAt = e.instant, updatedBy = e.subject)
+      // Deprecating a cross-project resolver
+      case (c: CrossProjectCurrent, e: ResolverDeprecated) if !c.deprecated =>
+        c.copy(rev = e.rev, deprecated = true, updatedAt = e.instant, updatedBy = e.subject)
+
+      // Cases that should happen and that should have been filtered out by the evaluation
+      case (s, _)                                                           => s
+    }
+  }
+
+  private[delta] def evaluate(
+      isOpen: ProjectRef => IO[ReadOnlyProject, Unit]
+  )(state: ResolverState, command: ResolverCommand)(implicit
+      caller: Caller,
+      clock: Clock[UIO]
+  ): IO[ResolverRejection, ResolverEvent] = {
+
+    // All provided identities must be part of the caller identities
+    def validateIdentities(identities: Set[Identity]): IO[ResolverRejection, Unit] = {
+      if (identities.isEmpty)
+        IO.raiseError(NoIdentities)
+      else {
+        val missing = identities.diff(caller.identities)
+        if (missing.isEmpty) {
+          IO.unit
+        } else {
+          IO.raiseError(InvalidIdentities(missing))
+        }
+      }
+    }
+
+    val stateMachine = (state, command) match {
```

Comment:
I think that having separate states, commands and events for variations of resolvers is making things much more complicated...

Reply:
If it is rewritten with a single state/command/event with a discriminator field, the complexity will appear anyway via this field no ?
And unused fields will appear for in-project resolvers.